### ISSUE TITLE
ASPIRE Logging 

### ISSUE THREAD

**Issue by garrettwrong** [ComputationalCryoEM/ASPIRE-Python#212]
ASPIRE should probably have a single dedicated logging module which wraps and configures the python logger in a consistent way.  This would be imported throughout the code in place of the python logger, and should avoid a lot of redundancy.  Probably I have some similar codes in my toolbox I can re-use to help us with this.

Some thoughts:

- Global configuration of print style/template (do we include module/file, times, etc)
- Perhaps extend the Debug, Info, Warning, Error with some extra tiers.
- Allow users to save these diagnostic logs to files (the python logger has some standard ways to tee the stream).
- Basically configurable by users though `config.ini`. (logging level, optional file names etc.)
- It might make sense to have a function in the logger module that will save the `config.ini` and some system/software information (release, if a repo, is it dirty(modified) and what is the hash then). This is both for scientific curation and potential debugging.
- Whatever paradigms we pick here need to be rationalized between things like the tutorial scripts, notebooks, and an interface (cli or ?).  We can probably design something that "should just work" for all of these things now.

**Comment by junchaoxia** [comment]
As far as I know, the  Basis, Cov3d, Cov2d, Orientation Estimation classes and corresponding stand alone commend in the new framework are using the Python logger module. We probably should try to use the same logger module for future development first. Of course the detailed configuration should be discussed an finalized  before the first formal release.

**Comment by garrettwrong** [comment]
When I wrote this issue, I had determined there are no less than three logging approaches in the repo's code (all building on the standard library logger).  I have not exhaustively searched for more.  The issue is simply to make one single approach/module, use that specifically instead of having each file/PR/module quasi-duplicate code, yield different configurations, and lack major features being consistent (like level/dumping to files).  The standard python logger is very powerful/flexible, so usually projects centralize it in a meaningful way. I think I've made some main features to aim for pretty clear. I am _not_ saying that some of the approaches are not able/doing some of these things, that isn't the problem...

Considering how we are working at this time, this issue is probably "that discussion". If you have specific contributions, or additional features, this is the best place to note them.  Otherwise I am glad it is now obvious to you this should be done.

The main (logging) things we don't do well right now are consistency, file streams, and logging level configuration.  Much of the remaining items can be extracted from the approaches throughout the code

**Comment by janden** [comment]
This makes sense. I remember seeing some logging boilerplate across the different modules that I assume would be eliminated by this. It would be good to standardize it more generally as well, though.

**Comment by garrettwrong** [comment]
One thing that I noticed recently is that the logging messages to the terminal don't delineate the level in their current configuration. At least not for some common modules (maybe for some other, harder to say rn).

Also under `pytest -s` we get duplicates, which I think can be fixed, since I haven't had that problem with other Python projects.

I'm hoping this brings enough value when you support users that it merits the work.

**Comment by janden** [comment]
When you say level, are you referring to Debug/Info/Warning/etc. or to some nesting counter (that is, higher-level code would output logging info at a certain level, which lower-level code would log at a different one)? Sorry just not familiar with the terminology here.

**Comment by garrettwrong** [comment]
The former. I shameless stole the example from their doc page:

```
import logging
logging.basicConfig(filename='example.log',level=logging.DEBUG)
logging.debug('This message should go to the log file')
logging.info('So should this')
logging.warning('And this, too')
```
```
DEBUG:root:This message should go to the log file
INFO:root:So should this
WARNING:root:And this, too
```

**Comment by garrettwrong** [comment]
Right now we just get `timestamp msg`

**Comment by vineetbansal** [comment]
Isn't individual modules getting a logger by `__name__` and emitting records to it sufficient? It's hierarchical and lightweight. We can always configure formats and what to do with the logs in a central place (right now it's done in `__init__.py` I think, with a `dictConfig` instead of `basicConfig` done through `config.ini` or `config.json`). I would recommend that approach instead of getting a centralized logger object through an import.

We could have imports on non-standard logging-levels for modules (e.g. `LOG_LEVEL_CRYO_OPERATOR=32` etc.) to use aside from the standard log levels.

There's the issue of what individual modules deem log-worthy, and at what level. That is perhaps all over the place right now.

**Comment by garrettwrong** [comment]
Closed by #288 .
